bpf/lib/classifiers.h

```diff
@@ -16,26 +19,147 @@ enum {
 	CLS_FLAG_IPV6	   = (1 << 0),
 	/* Packet originates from a L3 device (no ethernet header). */
 	CLS_FLAG_L3_DEV    = (1 << 1),
+	/* Packet uses underlay VXLAN. */
+	CLS_FLAG_VXLAN     = (1 << 2),
+	/* Packet uses underlay Geneve. */
+	CLS_FLAG_GENEVE    = (1 << 3),
 };
 
 /* Wrapper for specifying empty flags during the trace/drop event. */
 #define CLS_FLAG_NONE ((cls_flags_t)0)
 
+/* Overlay traffic should be observed from bpf_xdp, bpf_host, and bpf_wireguard. */
+#if defined(HAVE_ENCAP) && \
+	(defined(IS_BPF_XDP) || defined(IS_BPF_HOST) || defined(IS_BPF_WIREGUARD))
+# define OVERLAY_CLASSIFIERS
+#endif
+
 /**
  * ctx_classify
  * @ctx: socket buffer
  *
  * Compute classifiers (CLS_FLAG_*) for the given packet to be used during
- * trace/drop notification events.
+ * trace/drop notification events. In the worst case, three different checks
+ * are performed to obtain a sufficiently informative classification:
+ *
+ * 1. packets from L3 devices: needed to signal CLS_FLAG_L3_DEV and CLS_FLAG_IPV6.
+ * 2. ctx->mark: in case it carries informative magic values (SKB only).
+ * 3. packet headers: to look for known traffic patterns (ex. UDP+OverlayPort).
+ *
+ * The check (1) is always performed, (2) is skipped in XDP but executed in other
+ * contexts, and (3) runs only when (2) is not informative enough.
+ *
+ * NOTE: this function uses `revalidate_data` rather than `revalidate_data_pull`
+ * to preserve verifier complexity. If (3) is expected/required to run, make sure
+ * to use this when `revalidate_data_pull` has already been invoked.
  */
 static __always_inline cls_flags_t
-ctx_classify(const struct __ctx_buff *ctx __maybe_unused)
+ctx_classify(struct __ctx_buff *ctx)
```

Comment:
Did you consider whether the caller can already pass the ethertype here?

For `send_trace_notify()` this would be just `ctx_get_protocol()`. But `send_trace_notify4()` and `send_trace_notify6()` could optimize and pass `ETH_P_IP*` straight away.

Reply:
Just out of curiosity - is that on-top of the recent [bump to LLVM 19](https://github.com/cilium/cilium/pull/39632)?